File: csharp/Resources/McpResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// MCP runtime and trust-store Wave 3 operations
    /// </summary>
    public class McpResource
    {
        private readonly AuthdogClient _client;

        public McpResource(AuthdogClient client)
        {
            _client = client;
        }

        private string? Runtime(string? token) => token ?? _client.EnvironmentSecret;

        public Task<JObject> IngestEventsAsync(object body, string? token = null) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                "/v1/mcp/events",
                body,
                accessToken: Runtime(token));

        public JObject IngestEvents(object body, string? token = null) =>
            IngestEventsAsync(body, token).GetAwaiter().GetResult();

        public Task<JObject> ResolveAsync(string subject, string? token = null) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                "/v1/mcp/trust-store/resolve",
                query: AuthdogClient.Params(("subject", subject)),
                accessToken: Runtime(token));

        public JObject Resolve(string subject, string? token = null) =>
            ResolveAsync(subject, token).GetAwaiter().GetResult();

        public Task<JObject> ListEntriesAsync(string tenantId, string environmentId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/mcp/trust-store");

        public JObject ListEntries(string tenantId, string environmentId) =>
            ListEntriesAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> CreateEntryAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/mcp/trust-store",
                body);

        public JObject CreateEntry(string tenantId, string environmentId, object body) =>
            CreateEntryAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> GetEntryAsync(string tenantId, string environmentId, string entryId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/mcp/trust-store/{entryId}");

        public JObject GetEntry(string tenantId, string environmentId, string entryId) =>
            GetEntryAsync(tenantId, environmentId, entryId).GetAwaiter().GetResult();

        public Task<JObject> UpdateEntryAsync(string tenantId, string environmentId, string entryId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Patch,
                $"{AuthdogClient.Env(tenantId, environmentId)}/mcp/trust-store/{entryId}",
                body);

        public JObject UpdateEntry(string tenantId, string environmentId, string entryId, object body) =>
            UpdateEntryAsync(tenantId, environmentId, entryId, body).GetAwaiter().GetResult();

        public Task<JObject> DeleteEntryAsync(string tenantId, string environmentId, string entryId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Delete,
                $"{AuthdogClient.Env(tenantId, environmentId)}/mcp/trust-store/{entryId}");

        public JObject DeleteEntry(string tenantId, string environmentId, string entryId) =>
            DeleteEntryAsync(tenantId, environmentId, entryId).GetAwaiter().GetResult();

        public Task<JObject> AddKeyAsync(string tenantId, string environmentId, string entryId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/mcp/trust-store/{entryId}/keys",
                body);

        public JObject AddKey(string tenantId, string environmentId, string entryId, object body) =>
            AddKeyAsync(tenantId, environmentId, entryId, body).GetAwaiter().GetResult();

        public Task<JObject> RevokeKeyAsync(string tenantId, string environmentId, string entryId, string keyId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Delete,
                $"{AuthdogClient.Env(tenantId, environmentId)}/mcp/trust-store/{entryId}/keys/{keyId}");

        public JObject RevokeKey(string tenantId, string environmentId, string entryId, string keyId) =>
            RevokeKeyAsync(tenantId, environmentId, entryId, keyId).GetAwaiter().GetResult();

        public Task<JObject> RotateKeyAsync(
            string tenantId,
            string environmentId,
            string entryId,
            string keyId,
            object? body = null) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/mcp/trust-store/{entryId}/keys/{keyId}/rotate",
                body);

        public JObject RotateKey(
            string tenantId,
            string environmentId,
            string entryId,
            string keyId,
            object? body = null) =>
            RotateKeyAsync(tenantId, environmentId, entryId, keyId, body).GetAwaiter().GetResult();

        public Task<JObject> RevokeEntryAsync(string tenantId, string environmentId, string entryId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/mcp/trust-store/{entryId}/revoke");

        public JObject RevokeEntry(string tenantId, string environmentId, string entryId) =>
            RevokeEntryAsync(tenantId, environmentId, entryId).GetAwaiter().GetResult();

        public Task<JObject> VerifyEntryAsync(string tenantId, string environmentId, string entryId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/mcp/trust-store/{entryId}/verify",
                body);

        public JObject VerifyEntry(string tenantId, string environmentId, string entryId, object body) =>
            VerifyEntryAsync(tenantId, environmentId, entryId, body).GetAwaiter().GetResult();
    }
}

File: java/src/main/java/com/authdog/resources/McpResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.fasterxml.jackson.databind.JsonNode;
import java.util.Map;

/**
 * MCP runtime and trust-store operations.
 */
public final class McpResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create an MCP helper.
     * @param clientParam Authdog client
     */
    public McpResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * Ingest MCP runtime events.
     * @param body request body
     * @return ingest envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode ingestEvents(final Object body)
            throws AuthenticationException, ApiException {
        return ingestEvents(body, null);
    }

    /**
     * Ingest MCP runtime events with a token override.
     * @param body request body
     * @param token optional environment secret override
     * @return ingest envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode ingestEvents(final Object body, final String token)
            throws AuthenticationException, ApiException {
        return client.request("POST", "/v1/mcp/events", body, null,
                JsonNode.class, runtimeToken(token));
    }

    /**
     * Resolve a trust-store subject at runtime.
     * @param subject subject identifier
     * @return resolve envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode resolve(final String subject)
            throws AuthenticationException, ApiException {
        return resolve(subject, null);
    }

    /**
     * Resolve a trust-store subject with a token override.
     * @param subject subject identifier
     * @param token optional environment secret override
     * @return resolve envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode resolve(final String subject, final String token)
            throws AuthenticationException, ApiException {
        final Map<String, String> params = QueryParams.create();
        QueryParams.put(params, "subject", subject);
        return client.request("GET", "/v1/mcp/trust-store/resolve",
                null, QueryParams.orNull(params), JsonNode.class,
                runtimeToken(token));
    }

    /**
     * List trust-store entries for an environment.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return entries envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listEntries(final String tenantId,
                                final String environmentId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/mcp/trust-store",
                null, null, JsonNode.class);
    }

    /**
     * Create a trust-store entry.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return created entry envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode createEntry(final String tenantId,
                                final String environmentId,
                                final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/mcp/trust-store",
                body, null, JsonNode.class);
    }

    /**
     * Get a trust-store entry.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param entryId entry ID
     * @return entry envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode getEntry(final String tenantId,
                             final String environmentId,
                             final String entryId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/mcp/trust-store/" + entryId,
                null, null, JsonNode.class);
    }

    /**
     * Update a trust-store entry.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param entryId entry ID
     * @param body request body
     * @return updated entry envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode updateEntry(final String tenantId,
                                final String environmentId,
                                final String entryId,
                                final Object body)
            throws AuthenticationException, ApiException {
        return client.request("PATCH",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/mcp/trust-store/" + entryId,
                body, null, JsonNode.class);
    }

    /**
     * Delete a trust-store entry.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param entryId entry ID
     * @return delete envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode deleteEntry(final String tenantId,
                                final String environmentId,
                                final String entryId)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/mcp/trust-store/" + entryId,
                null, null, JsonNode.class);
    }

    /**
     * Add a key to a trust-store entry.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param entryId entry ID
     * @param body request body
     * @return key envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode addKey(final String tenantId,
                           final String environmentId,
                           final String entryId,
                           final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/mcp/trust-store/" + entryId + "/keys",
                body, null, JsonNode.class);
    }

    /**
     * Revoke a key on a trust-store entry.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param entryId entry ID
     * @param keyId key ID
     * @return revoke envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode revokeKey(final String tenantId,
                              final String environmentId,
                              final String entryId,
                              final String keyId)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/mcp/trust-store/" + entryId + "/keys/"
                        + keyId,
                null, null, JsonNode.class);
    }

    /**
     * Rotate a key on a trust-store entry.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param entryId entry ID
     * @param keyId key ID
     * @return rotate envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode rotateKey(final String tenantId,
                              final String environmentId,
                              final String entryId,
                              final String keyId)
            throws AuthenticationException, ApiException {
        return rotateKey(tenantId, environmentId, entryId, keyId, null);
    }

    /**
     * Rotate a key on a trust-store entry with an optional body.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param entryId entry ID
     * @param keyId key ID
     * @param body optional request body
     * @return rotate envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode rotateKey(final String tenantId,
                              final String environmentId,
                              final String entryId,
                              final String keyId,
                              final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/mcp/trust-store/" + entryId + "/keys/"
                        + keyId + "/rotate",
                body, null, JsonNode.class);
    }

    /**
     * Revoke a trust-store entry.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param entryId entry ID
     * @return revoke envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode revokeEntry(final String tenantId,
                                final String environmentId,
                                final String entryId)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/mcp/trust-store/" + entryId + "/revoke",
                null, null, JsonNode.class);
    }

    /**
     * Verify a trust-store entry.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param entryId entry ID
     * @param body request body
     * @return verify envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode verifyEntry(final String tenantId,
                                final String environmentId,
                                final String entryId,
                                final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/mcp/trust-store/" + entryId + "/verify",
                body, null, JsonNode.class);
    }

    /**
     * Resolve the MCP runtime Bearer token.
     * @param token optional override
     * @return token or environment secret
     */
    private String runtimeToken(final String token) {
        return token != null ? token : client.getEnvironmentSecret();
    }
}
